src/jrd/Attachment.cpp

```diff
@@ -111,10 +111,11 @@ void Jrd::Attachment::destroy(Attachment* const attachment)
 }
 
 
-MemoryPool* Jrd::Attachment::createPool()
+MemoryPool* Jrd::Attachment::createPool(MemoryPool* parent)
 {
-	MemoryPool* const pool = MemoryPool::createPool(att_pool, att_memory_stats);
-	att_pools.add(pool);
+	MemoryPool* const pool = MemoryPool::createPool(parent ? parent : att_pool, att_memory_stats);
+	if (!parent)
+		att_pools.add(pool);
 	return pool;
 }
 
```

Comment:
Why is the parent-based new pool not accounted in att_pools? It's still a proper pool and must be destroyed via deletePool() in ~Attachment() if not deleted explicitly.

Reply:
When the scratch pool is manually deleted (DML statement compiled without error) the thing works with or without registering it in att_pools, as deletePool will be called and remove it.

But for the non-DML case or DML with exception in compilation, that means only the statement pool is explicitly destroyed (scratch is child of statement). In this case, it will try to destroy an already deallocated pool in detach.

Is there any real problem in deleting a child pool via its parent?

Reply:
After some tests I see that current usage (let child pool die with parent) does not decrement child pool's memory statistics when parent dies.

I think parent pools should maintain a list of its active children and when parent is destroyed it should decrement statistics from children.

Reply:
@asfernandes Does child pool unmap memory got from OS when parent is destroyed? As far as I remember we always used to explicitly destroy child pools before parent.

Reply:
> Is there any real problem in deleting a child pool via its parent?
AFAIK, it just never happened in our code, so I cannot answer for sure. "Parent" concept was used to redirect allocations up to some amount, but the child pool may have its own memory blocks that the parent is not aware of (and thus cannot deallocate).

Reply:
@AlexPeshkoff if I manually destroy parent then follow and destroy child, a segfault happens and this seems very correct, as should be the concept of child pool.

But if I only destroy parent, child does not decrement its usage in the memory stats object.

So IMO parent should do this for its active children when parent is deleted.

Reply:
When a child pool is going to allocate large amounts of memory, are these blocks requested from parent or are mapped directly with the OS?

Reply:
> But if I only destroy parent, child does not decrement its usage in the memory stats object.

AFAIU, the idea is that the parent should never die before any of its children. Once they're dead, it makes no sense to adjust their counters.

Reply:
In other words, "delete by pool" concept does not apply to child pools. And I'm not sure it should - it would complicate things without any real gain.

Reply:
Then how should you prefer? Making the scratch pool independent (not as child of statement) or use parent-child implementing some logic related to att_pools to never delete child after parent?

Reply:
> When a child pool is going to allocate large amounts of memory, are these blocks requested from parent or are mapped directly with the OS?

They are mapped directly from OS.

Reply:
I'd prefer to make the scratch pool independent but please ensure that it dies when necessary (prepare failure?).

Reply:
@dyemanov is it ok to merge in master now?